### PR TITLE
Homework4 ncast01

### PR DESCRIPTION
**Pull request by ncast01**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. However, there are a few areas where it could be improved. Here are my suggestions:

1) **Coverage of the homework prompt:** Your solution covers most aspects of the prompt, but it seems to be missing the creation of the `players_state_tracking` table. This table is important for storing the state change tracking information for each player. You might want to consider adding a `CREATE TABLE` statement at the beginning of your solution, similar to the example solution.

2) **Readability:** Your code is generally well-formatted and easy to read. However, it could benefit from more comments explaining your thought process, especially in the more complex parts of your code. For example, you could add comments explaining the logic behind your `CASE` statements and how they relate to the different player states.

3) **Robustness and error handling:** Your code does not seem to handle the case where a player's `is_active` status is `NULL`. This could potentially lead to errors or incorrect results. You might want to add some error handling or checks to ensure that `is_active` is not `NULL` before using it in your `CASE` statements.

4) **Modularity and reusability:** Your code is structured in a way that makes it relatively easy to understand and reuse specific components. However, the `combined` CTE could potentially be broken down into smaller, more focused CTEs to improve readability and reusability. For example, you could have separate CTEs for calculating the `first_active_season` and `last_active_season` for each player.

5) **Logic and coherence:** Your logic for determining the `season_active_state` seems to be slightly off. According to the prompt, a player is considered 'New' if they are entering the league, 'Retired' if they are leaving the league, 'Continued Playing' if they are staying in the league, 'Returned from Retirement' if they are coming out of retirement, and 'Stayed Retired' if they stay out of the league. However, your code seems to consider a player 'New' if their `first_active_season` and `last_active_season` are the same, which might not always be the case for a new player. You might want to revise your logic to more accurately reflect the conditions in the prompt.

Keep up the good work, and don't hesitate to ask if you have any questions about these suggestions!

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

File: submission/query_1.sql [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+FROM last_season ls 
+FULL OUTER JOIN current_season cs
+ON ls.player_name = cs.player_name
+
+)
+
+
+SELECT
+
```
Comment (HjersmanJ) [inline submission/query_1.sql:57]: Unusual to have this kind of white space in SQL files. Would be best to keep these code blocks contiguous.

File: submission/query_2.sql [hunk 2]
```diff
@@ -0,0 +1,45 @@
+-- Write a query (query_2) that uses GROUPING SETS to perform aggregations of the nba_game_details data. Create slices that aggregate along the following combinations of dimensions:
+
+-- player and team
+-- player and season
+-- team
```
Comment (HjersmanJ) [inline submission/query_2.sql:5]: Would be ideal to have this kind of explanation either summarized here or explained in the README.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. However, there are a few areas where it could be improved. Here are my suggestions:

1) **Coverage of the homework prompt:** Your solution covers most aspects of the prompt, but it seems to be missing the creation of the `players_state_tracking` table. The example solution starts by creating this table, which is then used to store the state change tracking information. You might want to consider adding this to your solution.

2) **Logical flow:** Your use of Common Table Expressions (CTEs) to break down the problem into smaller, more manageable parts is commendable. However, the `current_season` CTE seems to be missing a filter for `is_active` players. This might lead to incorrect results, as it would include players who are not currently active.

3) **Readability:** Your code is generally well-formatted and easy to read. However, it could benefit from more comments explaining your thought process, especially in the more complex parts of the code. For example, the logic behind the `CASE` statements in the `combined` CTE could be explained in more detail.

4) **Robustness and error handling:** Your code does not seem to handle potential errors or edge cases. For example, what happens if the `player_name` is `NULL` or if the `season` is not a valid year? Adding some error handling or input validation could make your code more robust.

5) **Modularity and reusability:** Your use of CTEs makes your code more modular and easier to understand. However, the `combined` CTE is quite complex and might be difficult to reuse in other contexts. You might want to consider breaking it down into smaller parts.

In conclusion, your solution is on the right track, but could benefit from some improvements. Keep up the good work!

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

File: submission/query_2.sql [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  SUM(pf) AS pt,
+  SUM(reb) AS reb,
+  SUM(ast) AS ast
+  
+  FROM base
+
+  GROUP BY GROUPING SETS(
+       (player_name, team),
+       (player_name, season),
+       (team)
```
Comment (HjersmanJ) [inline submission/query_2.sql:44]: You need to alias the grouping sets with the GROUPING() function. Otherwise this makes them harder to use then they are designed for.

File: submission/query_3.sql [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+player_name,
+team,
+pts
+
+FROM aggregations
+WHERE season = 'Overall' -- getting total points across all seasons
+AND player_name != 'Overall'
+
+ORDER BY pts DESC
+LIMIT 1
```
Comment (HjersmanJ) [inline submission/query_3.sql:56]: This code is fine. However, because you didn't create the grouping sets to easily reference, you had to include the code above, which is unnecessary. The intent is to simply call the grouping set with the desired query at the aggregation level.

**Review by HjersmanJ — APPROVED** [review]
Your outputs are fine. However, your usage of grouping sets is not quite as intended. 

They are designed for frequently used analytics that can be easily accessed by analysts. If they are not instantiated properly, they are more difficult to use for your downstream consumers. They are intended to be defined with supporting well-written documentation (such as the README).

Regardless, you have completed the queries to a satisfactory degree and pass this assignment. Good job.
